File: lc/go/valid-palindrome.go
// i         j
// A   m a q , a 
// 0 1 2 3 4 5 6
//     l
//         r
// sl sr a a 
// implmenet a revert but always continue until 
// 
func isPalindrome(s string) bool {
    for l, r := 0, len(s) - 1; l < r; {
        sl, sr := s[l], s[r]
        if ('A' <= sl && sl <= 'Z') { sl += ('a'-'A')}
        if ('A' <= sr && sr <= 'Z') { sr += ('a'-'A')}
        if ! (('a' <= sl && sl <= 'z') || ('0' <= sl && sl <= '9') ) { 
            l++
            continue
        }
        if ! (('a' <= sr && sr <= 'z') || ('0' <= sr && sr <= '9')) { 
            r--
            continue
        }
        if l > r || sl != sr { return false }
        l, r = l + 1, r - 1
    }
    return true
}

File: lc/py/valid-palindrome.py
# a jfdkj;l;fsd;;;
# i
#                j
# i moves until next alphanum
# j does the same other direction
# if they have not hit the same char
#     compare for palindronme --> false if not the same
# return true if they are at same char
# ;;aa
#    i  
#   j
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = s.lower()
        n = len(s)
        i,j = 0, n-1
        while i<j :
            while i<j and not s[i].isalnum():
                i += 1
            while i<j and not s[j].isalnum():
                j -= 1
            if s[i] != s[j]:
                return False
            i,j = i+1, j-1
        return True # won't happen